sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala

```diff
@@ -22,10 +22,40 @@ import java.time.chrono.IsoChronology
 import java.time.format.{DateTimeFormatter, DateTimeFormatterBuilder, ResolverStyle}
 import java.time.temporal.{ChronoField, TemporalAccessor, TemporalQueries}
 import java.util.Locale
+import java.util.concurrent.ConcurrentHashMap
 
 trait DateTimeFormatterHelper {
+  protected def toInstantWithZoneId(temporalAccessor: TemporalAccessor, zoneId: ZoneId): Instant = {
+    val localTime = if (temporalAccessor.query(TemporalQueries.localTime) == null) {
+      LocalTime.ofNanoOfDay(0)
+    } else {
+      LocalTime.from(temporalAccessor)
+    }
+    val localDate = LocalDate.from(temporalAccessor)
+    val localDateTime = LocalDateTime.of(localDate, localTime)
+    val zonedDateTime = ZonedDateTime.of(localDateTime, zoneId)
+    Instant.from(zonedDateTime)
+  }
+}
+
+object DateTimeFormatterHelper {
+  private val cache = new ConcurrentHashMap[(String, Locale), DateTimeFormatter]()
+
+  def getFormatter(pattern: String, locale: Locale): DateTimeFormatter = {
+    val key = (pattern, locale)
+    var formatter = cache.get(key)
```

Comment:
`computeIfAbsent` ought to be simpler and more efficient here, to compute the value only if it isn't already present

Reply:
but it blocks other threads during lambda computation: `...Some attempted update operations on this map by other threads may be blocked while computation is in progress, so the computation should be short and simple...` but current implementation does not.

Reply:
Sure, but that's necessary to avoid computing it more than once right? and only is an issue if multiple threads need the value at once, the first time. if it's blocking for milliseconds that seems OK. It would be an issue if it meant every subsequent access slowed down or was unnecessarily contended.

Reply:
I just follow implementation of FastDateFormat from Apache Commons lang3: https://github.com/apache/commons-lang/blob/8e8b8e05e4eb9aa009444c2fea3552d28b57aa98/src/main/java/org/apache/commons/lang3/time/FormatCache.java#L71-L91

Reply:
Commons lang3 supports Java 7, so couldn't use computeIfAbsent. I presume it would if it could. I don't feel super strongly about it, but think we can take advantage of Java 8 here. It saves a second lookup, and in so doing, avoids the (fairly harmless) race condition here -- multiple threads can find the instance isn't cached and compute and try to put the result. It is still correct but not optimal.

Reply:
I agree with @srowen. If the blocking frequently occurs (i.e. key does not exist), this cache does not work effectively.  
If a key frequently exists, the blocking will not occur frequently.